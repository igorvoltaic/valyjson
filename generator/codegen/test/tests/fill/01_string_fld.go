package testo

//valyjson:optional
type TestStr01 struct {
	Field string `json:"field"`
}
