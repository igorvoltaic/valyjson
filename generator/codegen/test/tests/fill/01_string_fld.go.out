// Code generated [github.com/iv-menshenin/valyjson]; DO NOT EDIT.
package testo

import (
	"bytes"
	"fmt"
	"strconv"
	"github.com/valyala/fastjson"
)
// jsonParserTestStr01used for pooling Parsers for TestStr01 JSONs.
var jsonParserTestStr01 fastjson.ParserPool

// UnmarshalJSON implements json.Unmarshaler
func (s *TestStr01) UnmarshalJSON(data []byte) error {
	parser := jsonParserTestStr01.Get()
	// parses data containing JSON
	v, err := parser.ParseBytes(data)
	if err != nil {
		return err
	}
	defer jsonParserTestStr01.Put(parser)
	return s.FillFromJson(v, "")
}

// FillFromJson recursively fills the fields with fastjson.Value
func (s *TestStr01) FillFromJson(v *fastjson.Value, objPath string) (err error) {
	if field := v.Get("field"); field != nil {
		if field.Type() != fastjson.TypeString {
			err = fmt.Errorf("value doesn't contain string; it contains %s", field.Type())
			return fmt.Errorf("error parsing '%sfield' value: %w", objPath, err)
		}
		valField := field.String()
		s.Field = valField
	}
	return nil
}

// validate checks for correct data structure
func (s *TestStr01) validate(v *fastjson.Value, objPath string) error {
	o, err := v.Object()
	if err != nil {
		return err
	}
	o.Visit(func(key []byte, _ *fastjson.Value) {
		if err != nil {
			return
		}
		if bytes.Equal(key, []byte{'f', 'i', 'e', 'l', 'd'}) {
			return
		}
		if objPath == "" {
			err = fmt.Errorf("unexpected field '%s' in the root of the object", string(key))
		} else {
			err = fmt.Errorf("unexpected field '%s' in the '%s' path", string(key), objPath)
		}
	})
	return nil
}

// valueIsNotNull allows you to determine if the value is contained in a Json structure.
// Checks if the structure itself or the value is Null.
func valueIsNotNull(v *fastjson.Value) bool {
	return v != nil && v.Type() != fastjson.TypeNull
}
